rfcs/20190722-tflite-training.md

```diff
@@ -0,0 +1,444 @@
+usually required to train the model. Even if Tensor List ops aren't in the
+original inference graph, when constructing the training graph, Tensor List ops
+are often added automatically to memorize intermediate values, for computing
+gradients over the loop.
+
+Tensor Lists can also be used without loops but this rarely happens. In this
+case, Tensor List ops are required for inference and training.
+
+### Saving Format
+
+In TensorFlow, users can use `tf.Session` to train a model, and use `tf.train.Saver` to save the trained weights into TensorFlow checkpoint files.
+
+This proposal enables training with TensorFlow Lite interpreter (which is similar to `tf.Session`). In addition, we can provide utility classes (similar to `tf.train.Saver`) to save the trained format, or users can get the raw data of variables and save it by themselves.
+
+A few approaches to handle variable saving are described below. In the near term we will focus on the recommended approach. However, note that these approaches are not conflicting with each other, so it's possible to implement multiple of these.
+
+**Approach 1: Reuse TensorFlow CheckPoint format [recommended]**
+
+Pros:
+*   Interoperability: It's potentially easier to load a TensorFlow trained weights into TensorFlow Lite, and vice versa.
+*   Avoid designing yet another format.
+
+Cons:
+*   Currently the TensorFlow checkpoint parsing code is coupled with TensorFlow core. We may need to refactor and decouple with TensorFlow core, or rewrite the parser.
+
+**Approach 2: Write the variable value back to TensorFlow Lite model**
+
+Pros:
+*   Use the existing TensorFlow Lite format. Avoid introducing another format.
+
+Cons:
+*   Need to ship the FlatBuffer writer code into the binary.
+*   Have to write the entire model (which may contain other huge frozen constants) back to storage.
+*   Low interoperability with TensorFlow checkpoints
+
+**Approach 3: Implement TensorFlow Lite's own variable saving file format**
+
+Pros:
+*   This can be an extremely simple key-value mapping format
+
+Cons:
+*   Requires defining yet another format
+*   Low interoperability with TensorFlow checkpoints
+
+### Feeding data with tf.Example / tf.data.Dataset
+
+`tf.Example` and `tf.data.Dataset` are technically not training-specific
+requirements. These features can also be used in inference.
+
+The basic training can work by feeding the raw data (e.g. float values) into
+TensorFlow Lite interpreter. On top of this, we can support tf.Example and
+tf.data.Dataset to make this easier to use (likely via Flex runtime). There may be a
+performance drawback using these features.
+
+### Selective Registration
+
+Initially we will need to rely on Flex runtime to execute some of the fused
+gradient ops. This requires to add the gradient op kernels to Flex runtime, and
+it will bloat up the Flex binary size.
+
+We also plan to implement fused training ops as TensorFlow Lite builtin ops.
```

Comment:
Will there be *building configuration*, which controls whether to build the training-dedicated operators (or functionality) into binary? To me, there will be plenty of code to be added, resulting a larger tflite runtime.

Reply:
Would it be better if user can select what are the operators they need for their model? If users understand what they need they can better trim down the tflite binary size.

Reply:
How about the support of the Pixel Visual core during training? 
It seems that currently, Android Neural Network only has inference APIs.

Reply:
I'm assuming that you're commenting on this line. 

> We also plan to implement fused training ops as TensorFlow Lite builtin ops.

Yes. I think we need better selective registration mechanism on TensorFlow Lite builtin ops too, so the binary size stay small when training isn't used. 

However for the short term, we're just going to use the Flex runtime instead of defining builtin gradient ops.